### PR TITLE
[SandboxIR][Doc] Add a Doc file for Sandbox IR

### PR DESCRIPTION
**Pull request by vporpo**
This is under User Guides > Additional Topics > Sandbox IR.

### CODE REVIEW

**Comment by tschuett** [comment]
I don't mind if you don't want to link the RFC. A hint that will it will be the foundation for a new SLP vectorizer would be nice.

**Comment by vporpo** [comment]
Well, the IR itself is decoupled and not vectorizer-specific. We can link to the RFC once we start adding the Vector-specific extensions.

File: llvm/docs/SandboxIR.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+## Sandbox IR Value <-> LLVM IR Value Mapping
+LLVM IR and Sandbox IR are always in sync.
+
+- Forward mapping: Sandbox IR Value -> LLVM IR Value
+Each Sandbox IR Value contains an `llvm::Value *Val` member variable that points to the corresponding LLVM IR Value.
+
+- Reverse mapping: LLVM IR Value -> Sandbox IR Value
+This mapping is stored in `sandboxir::Context::LLVMValueToValue`.
+
+For example `sandboxir::User::getOperand(OpIdx)` for a `sandboxir::User * U` works as follows:
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:34]:
```suggestion
For example `sandboxir::User::getOperand(OpIdx)` for a `sandboxir::User *U` works as follows:
```

and below

File: llvm/docs/SandboxIR.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+## Sandbox IR is Write-Through
+Sandbox IR is designed to rely on LLVM IR for its state.
+So any change made to Sandbox IR objects directly updates the corresponding LLVM IR.
+
+This has the following benefits:
+- It minimizes the replication of state, and
+- It makes sure that Sandbox IR and LLVM IR are always in sync, which helps avoid bugs and removes the need for a lowering step.
+- No need for serialization/de-serialization infrastructure as we can rely on LLVM IR for it.
+
+The implementation is straight-forward:
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:48]:
```suggestion
The implementation is straightforward:
```

File: llvm/docs/SandboxIR.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+Sandbox IR API functions that modify the IR state call the corresponding LLVM IR function that modifies the LLVM IR's state.
+For example, for `sandboxir::User::setOperand(OpIdx, sandboxir::Value *Op)`:
+- We get the corresponding LLVM User: `llvm::User *LLVMU = cast<llvm::User>(Val)`
+- Next we get the corresponding LLVM Operand: `llvm::Value *LLVMOp = Op->Val`
+- Finally we modify `LLVMU`'s operand: `LLVMU->setOperand(OpIdx, LLVMOp)`.
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:53]:
```suggestion
- Finally we modify `LLVMU`'s operand: `LLVMU->setOperand(OpIdx, LLVMOp)`
```

File: llvm/docs/SandboxIR.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+     ...
+}
+```
+
+Sandbox IR is also using a similar type system with `isa<>, cast<>, dyn_cast<>`.
+
+# Design
+
+## Sandbox IR Value <-> LLVM IR Value Mapping
+LLVM IR and Sandbox IR are always in sync.
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:26]: this sentence seems out of place. either explain it more (maybe earlier), combine it with the `Sandbox IR is Write-Through` section

File: llvm/docs/UserGuides.rst [hunk 5]
```diff
@@ -288,3 +290,6 @@ Additional Topics
 
 :doc:`RISCV/RISCVVectorExtension`
    This document describes how the RISC-V Vector extension can be expressed in LLVM IR and how code is generated for it in the backend.
+
+:doc:`Sandbox IR <SandboxIR>`
+   Describes the design and usage of Sandbox IR, a transactional layer over LLVM IR.
```
Comment (aeubanks) [inline llvm/docs/UserGuides.rst:295]:
```suggestion
   This document describes the design and usage of Sandbox IR, a transactional layer over LLVM IR.
```
(for consistency with above)

File: llvm/docs/SandboxIR.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# Sandbox IR: A transactional layer over LLVM IR
+
+Sandbox IR is an IR layer on top of LLVM IR that allow you to save/restore its state.
+
+# API
+The Sandbox IR API is designed to make it feel like LLVM.
```
[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/docs/SandboxIR.md:6]: Done

File: llvm/docs/SandboxIR.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+So Sandbox IR replicates many common API classes and functions to mirror the LLVM API.
```
[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/docs/SandboxIR.md:7]: Done

File: llvm/docs/SandboxIR.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+The class hierarchy is quite similar (but in the `sandboxir` namespace), for example here is a small part of it:
```
[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/docs/SandboxIR.md:8]: Done

File: llvm/docs/SandboxIR.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+              /  \
+            User BasicBlock ...
+           /   \
+  Instruction Constant
+        /
+     ...
+}
+```
+
+Sandbox IR is also using a similar type system with `isa<>, cast<>, dyn_cast<>`.
```
[the remark this answers is not in the document]
Comment (vporpo) [reply llvm/docs/SandboxIR.md:21]: Done

File: llvm/docs/SandboxIR.md (hunk 1 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:34]: Done

File: llvm/docs/SandboxIR.md (hunk 2 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:48]: Done

File: llvm/docs/SandboxIR.md (hunk 3 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:53]: Done

File: llvm/docs/SandboxIR.md (hunk 4 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md]: I expanded on that a bit.

File: llvm/docs/UserGuides.rst (hunk 5 above)
Comment (vporpo) [reply llvm/docs/UserGuides.rst]: Done

File: llvm/docs/SandboxIR.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
+# Sandbox IR: A transactional layer over LLVM IR
+
+Sandbox IR is an IR layer on top of LLVM IR that allow you to save/restore its state.
```
Comment (slackito) [inline llvm/docs/SandboxIR.md:3]:
```suggestion
Sandbox IR is an IR layer on top of LLVM IR that allows you to save/restore its state.
```

**Review by slackito — APPROVED** [review]
LGTM. I assume further details on the save/restore mechanism will be added when the corresponding code lands, right?

**Comment by vporpo** [comment]
Yes, we will keep adding things to the doc as we are adding more functionality to the code.

File: llvm/docs/SandboxIR.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        /
+     ...
+}
+```
+
+# Design
+
+## Sandbox IR Value <-> LLVM IR Value Mapping
+Each LLVM IR Value maps to a single Sandbox IR Value.
+The reverse is also true in most cases, except for multi-Instruction Sandbox IR Instructions that may be defined in extensions the Sandbox IR.
```
Comment (tmsri) [inline llvm/docs/SandboxIR.md:24]: "except for multi-Instruction Sandbox IR Instructions that may be defined in extensions the Sandbox IR."

Could you rephrase?

Comment (vporpo) [reply llvm/docs/SandboxIR.md]: Yeah that was pretty bad :P I rephrased it. Though I'm not sure how to best differentiate the "base" SandboxIR and the future extensions.

File: llvm/docs/SandboxIR.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+
+- Forward mapping: Sandbox IR Value -> LLVM IR Value
+Each Sandbox IR Value contains an `llvm::Value *Val` member variable that points to the corresponding LLVM IR Value.
+
+- Reverse mapping: LLVM IR Value -> Sandbox IR Value
+This mapping is stored in `sandboxir::Context::LLVMValueToValue`.
+
+For example `sandboxir::User::getOperand(OpIdx)` for a `sandboxir::User *U` works as follows:
+- First we find the LLVM User: `llvm::User * LLVMU = U->Val`.
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:35]:
```suggestion
- First we find the LLVM User: `llvm::User *LLVMU = U->Val`.
```

File: llvm/docs/SandboxIR.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+## Sandbox IR is Write-Through
+Sandbox IR is designed to rely on LLVM IR for its state.
+So any change made to Sandbox IR objects directly updates the corresponding LLVM IR.
+
+This has the following benefits:
+- It minimizes the replication of state, and
+- It makes sure that Sandbox IR and LLVM IR are always in sync, which helps avoid bugs and removes the need for a lowering step.
+- No need for serialization/de-serialization infrastructure as we can rely on LLVM IR for it.
+
+The implementation is straightforward:
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:48]: maybe remove this line? it doesn't add very much and the colon implies the next line is the start of a list, which it isn't here

File: llvm/docs/SandboxIR.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+For example `sandboxir::User::getOperand(OpIdx)` for a `sandboxir::User *U` works as follows:
+- First we find the LLVM User: `llvm::User * LLVMU = U->Val`.
+- Next we get the LLVM Value operand: `llvm::Value *LLVMOp = LLVMU->getOperand(OpIdx)`
+- Finally we get the Sandbox IR operand that corresponds to `LLVMOp` by querying the map in the Sandbox IR context: `retrun Ctx.getValue(LLVMOp)`.
+
+## Sandbox IR is Write-Through
+Sandbox IR is designed to rely on LLVM IR for its state.
+So any change made to Sandbox IR objects directly updates the corresponding LLVM IR.
+
+This has the following benefits:
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:43]: do you want to mention that you can pass `llvm::Instruction`s to various cost modeling APIs? or is that too specific to put here?

File: llvm/docs/SandboxIR.md (hunk 11 above)
Comment (aeubanks) [reply llvm/docs/SandboxIR.md]: maybe add a separate section about extensions?

File: llvm/docs/SandboxIR.md (hunk 14 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:43]: Yeah I can add this point.

File: llvm/docs/SandboxIR.md (hunk 13 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:48]: Done

File: llvm/docs/SandboxIR.md (hunk 12 above)
Comment (vporpo) [reply llvm/docs/SandboxIR.md:35]: Done

File: llvm/docs/SandboxIR.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+# Sandbox IR: A transactional layer over LLVM IR
+
+Sandbox IR is an IR layer on top of LLVM IR that allows you to save/restore its state.
+
+# API
+The Sandbox IR API is designed to feel like LLVM, replicating many common API classes and functions to mirror the LLVM API.
+The class hierarchy is similar (but in the `llvm::sandboxir` namespace), for example here is a small part of it:
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:7]:
```suggestion
The class hierarchy is similar (but in the `llvm::sandboxir` namespace). For example here is a small part of it:
```

File: llvm/docs/SandboxIR.md [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+}
+```
+
+# Design
+
+## Sandbox IR Value <-> LLVM IR Value Mapping
+Each LLVM IR Value maps to a single Sandbox IR Value.
+The reverse is also true in most cases, except for Sandbox IR Instructions that map to more than one LLVM IR Instruction.
+Such instructions can be defined in extensions of the base Sandbox IR.
+
```
Comment (aeubanks) [inline llvm/docs/SandboxIR.md:26]: extra newline

**Review by aeubanks — APPROVED** [review]
lg with comments, thanks for writing this up
